core/src/main/scala/filodb.core/memstore/ratelimit/CardinalityTracker.scala

```diff
@@ -102,14 +114,104 @@ class CardinalityTracker(ref: DatasetRef,
+   * the cardinality count of 4 prefixes. They are -
+   * 1. (total across all ws)
+   * 2. (my_ws)
+   * 3. (my_ws, my_ns)
+   * 4. (my_ws, my_ns, my_metric)
+   *
+   * @param shardKey
+   */
+  def updateCardinalityCountsDS(shardKey: Seq[String]): Unit = {
+    (0 to shardKey.length).foreach { i =>
+
+      // update current prefix's cardinality count
+      val prefix = shardKey.take(i)
+      val cardCountRecord = cardinalityCountMapDS.get(prefix)
+        .map(x => (x._1 + 1, x._2))
+        .getOrElse((1, 0)) // child prefix update parent's childrenCount
+      cardinalityCountMapDS.put(prefix, cardCountRecord)
+
+      // update children count of parent and throw exception if quota reached
+      if (i > 0) {
+        val parentPrefix = shardKey.take(i - 1)
+
+        // we always add parent before the child, hence it is okay to get the parent prefix's record directly
+        // without the None check
+        val updatedCountRecord = cardinalityCountMapDS.get(parentPrefix)
+          .map(x => (x._1, x._2 + 1)).get
+
+        // check if number of children is higher than the given quota. This allows us guard our physical resources
+        // and avoid failures because of runaway cardinality
+        val childrenQuota = defaultChildrenQuota(parentPrefix.length)
+        if (updatedCountRecord._2 > childrenQuota) {
+          quotaExceededProtocol.quotaExceeded(ref, shard, prefix, childrenQuota)
+          throw QuotaReachedException(prefix, prefix, childrenQuota)
+        }
+
+        // store the updated parent's childrenCount
+        cardinalityCountMapDS.put(parentPrefix, updatedCountRecord)
+      }
+    }
+
+    if (cardinalityCountMapDS.size > dsCardinalityMapFlushCount) {
+      flushCardinalityDataToRocksDB()
+    }
+  }
+
+  /*
+  * Flush the cardinality data to RocksDB before reading the counts. The downsample cardinality count is built from
+  * scratch at a periodic interval and the caller of CardinalityTracker can also call this method to ensure all data
+  * is flushed to RocksDB
+  * */
+  def flushCardinalityDataToRocksDB(): Unit = {
+    if (cardinalityCountMapDS.size > 0) {
+      // iterate through map and store each prefix and count to the rocksDB
+      cardinalityCountMapDS.foreach(kv => {
+        modifyCountDS(kv._1, kv._2._1, kv._2._1, kv._2._2)
+      })
+      // clear the map
+      cardinalityCountMapDS.clear()
+    }
+  }
+
+  /**
+   * Used to store the cardinality count for the given prefix in the downsample cluster.
+   * NOTE: In downsample cluster, tsCount == activeTsCount. So totalDelta and activeDelta is same
+   * @param prefix usually contains labels _ws_, _ns_, _metric_ and different combinations of it
+   * @param totalDelta Increase in total timeseries
+   * @param activeDelta Increase in active timeseries
+   * @param childrenDelta Increase in children count
+   */
+  private def modifyCountDS(prefix: Seq[String],
```

Comment:
it seems the same key would be updated multiple time. And each update requires a read before it. This can cause read amplification and compaction in rocksdb.

If the performance is a real concern, I would suggest aggregate the value for the same key before put it into rocksdb.

Reply:
This method is only called by `updateCardinalityCountsDS` which aggregates the count before storing in the RocksDB :)

Reply:
While the performance when you tested is good, it's not guaranteed ordering of ``partKeys``  when we batch. So to @yu-shipit 's point you may see a key at least for a ws/ns after flush?  But not too worried about it at the same time,

Reply:
Yeah, the implementation doesn't make any assumption of ordering guarantees and same/keys prefix will most likely happen again. Chose this implementation to avoid any complexity and keep implementation plain and simple ( no sort or other transformation. The reason for threshold and guards here is to avoid runaway use of memory )